🐛 [realtime_database] nullable  type not handled properly  ( firebase_database: ^7.0.0-dev.1)

## Bug report

**Describe the bug**
Unhandled Exception: Null check operator used on a null value

### Error Log
![Screenshot 2021-05-03 at 8 03 57 AM](https://user-images.githubusercontent.com/32235964/116837179-f4a1bc00-abe6-11eb-8dd2-e8969581b8d3.png)

**getting above error in line no:42**

### issue with code  ( line no:36 and line no: 42)
![Screenshot 2021-05-03 at 8 05 14 AM](https://user-images.githubusercontent.com/32235964/116837238-2fa3ef80-abe7-11eb-97d5-1a00d87b291a.png)

```dart
  final res = await ref.runTransaction((mutableData) async {
      if (
          (mutableData.value as Map)["state"] == "none") {
        mutableData.value = {"state": "waiting"};
        return mutableData;
      } else {
        mutableData.value = {"state": "none"};
        return mutableData;
      }
    });
```

error log 
```
Reloaded 146 of 2375 libraries in 2,533ms.
E/MethodCallHandlerImpl( 4516): Unable to commit Snapshot update. Transaction failed.
E/MethodCallHandlerImpl( 4516): java.util.concurrent.ExecutionException: java.lang.Exception: Error code: error
E/MethodCallHandlerImpl( 4516): Error message: type 'Null' is not a subtype of type 'String'
E/MethodCallHandlerImpl( 4516): Error details: null
E/MethodCallHandlerImpl( 4516): 	at com.google.android.gms.tasks.Tasks.zza(com.google.android.gms:play-services-tasks@@17.1.0:69)
E/MethodCallHandlerImpl( 4516): 	at com.google.android.gms.tasks.Tasks.await(com.google.android.gms:play-services-tasks@@17.1.0:33)
E/MethodCallHandlerImpl( 4516): 	at io.flutter.plugins.firebase.database.MethodCallHandlerImpl$1.doTransaction(MethodCallHandlerImpl.java:433)
E/MethodCallHandlerImpl( 4516): 	at com.google.firebase.database.core.Repo.startTransaction(Repo.java:915)
E/MethodCallHandlerImpl( 4516): 	at com.google.firebase.database.DatabaseReference$4.run(DatabaseReference.java:492)
E/MethodCallHandlerImpl( 4516): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
E/MethodCallHandlerImpl( 4516): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/MethodCallHandlerImpl( 4516): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
E/MethodCallHandlerImpl( 4516): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/MethodCallHandlerImpl( 4516): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/MethodCallHandlerImpl( 4516): 	at java.lang.Thread.run(Thread.java:919)
E/MethodCallHandlerImpl( 4516): Caused by: java.lang.Exception: Error code: error
E/MethodCallHandlerImpl( 4516): Error message: type 'Null' is not a subtype of type 'String'
E/MethodCallHandlerImpl( 4516): Error details: null
E/MethodCallHandlerImpl( 4516): 	at io.flutter.plugins.firebase.database.MethodCallHandlerImpl$1$1$1.error(MethodCallHandlerImpl.java:417)
E/MethodCallHandlerImpl( 4516): 	at io.flutter.plugin.common.MethodChannel$IncomingResultHandler.reply(MethodChannel.java:212)
E/MethodCallHandlerImpl( 4516): 	at io.flutter.embedding.engine.dart.DartMessenger.handlePlatformMessageResponse(DartMessenger.java:105)
E/MethodCallHandlerImpl( 4516): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessageResponse(FlutterJNI.java:829)
E/MethodCallHandlerImpl( 4516): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodCallHandlerImpl( 4516): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MethodCallHandlerImpl( 4516): 	at android.os.Looper.loop(Looper.java:174)
E/MethodCallHandlerImpl( 4516): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/MethodCallHandlerImpl( 4516): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodCallHandlerImpl( 4516): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/MethodCallHandlerImpl( 4516): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
E/flutter ( 4516): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: Null check operator used on a null value
[38;5;244mE/flutter ( 4516): #0      new DataSnapshot._fromJson[39;49m
[38;5;244mE/flutter ( 4516): #1      DatabaseReference.runTransaction.toTransactionResult[39;49m
[38;5;244mE/flutter ( 4516): #2      DatabaseReference.runTransaction.<anonymous closure>[39;49m
[38;5;244mE/flutter ( 4516): #3      _rootRunUnary (dart:async/zone.dart:1362:47)[39;49m
[38;5;244mE/flutter ( 4516): #4      _CustomZone.runUnary (dart:async/zone.dart:1265:19)[39;49m
[38;5;244mE/flutter ( 4516): #5      _FutureListener.handleValue (dart:async/future_impl.dart:152:18)[39;49m
[38;5;244mE/flutter ( 4516): #6      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:704:45)[39;49m
[38;5;244mE/flutter ( 4516): #7      Future._propagateToListeners (dart:async/future_impl.dart:733:32)[39;49m
[38;5;244mE/flutter ( 4516): #8      Future._completeWithValue (dart:async/future_impl.dart:539:5)[39;49m
[38;5;244mE/flutter ( 4516): #9      _completeOnAsyncReturn (dart:async-patch/async_patch.dart:254:13)[39;49m
[38;5;244mE/flutter ( 4516): #10     MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart)[39;49m
E/flutter ( 4516): <asynchronous suspension>
E/flutter ( 4516):
```

using firebase_database: ^7.0.0-dev.2

--- a/packages/firebase_database/firebase_database/example/pubspec.yaml
+++ b/packages/firebase_database/firebase_database/example/pubspec.yaml
@@ -22,7 +22,7 @@ dependency_overrides:
     path: ../../../firebase_core/firebase_core_web
 
 dev_dependencies:
-  drive: 0.1.0
+  drive: 1.0.0-1.0.nullsafety.0
   flutter_driver:
     sdk: flutter
   test: any

--- a/packages/firebase_database/firebase_database/example/test_driver/firebase_database_e2e.dart
+++ b/packages/firebase_database/firebase_database/example/test_driver/firebase_database_e2e.dart
@@ -1,4 +1,3 @@
-// @dart = 2.9
 import 'package:drive/drive.dart' as drive;
 import 'package:firebase_core/firebase_core.dart';
 import 'package:firebase_database/firebase_database.dart';
@@ -23,7 +22,7 @@ void testsMain() {
       const String orderTestPath = 'ordered/';
 
       await Future.wait(testDocuments.map((map) {
-        String child = map['ref'] as String;
+        String child = map['ref']! as String;
         return database.reference().child('$orderTestPath/$child').set(map);
       }));
     });
@@ -36,7 +35,7 @@ void testsMain() {
           .onValue
           .first;
 
-      final ordered = testDocuments.map((doc) => doc['value'] as int).toList();
+      final ordered = testDocuments.map((doc) => doc['value']! as int).toList();
       ordered.sort();
 
       final documents = event.snapshot.value.values
@@ -60,7 +59,24 @@ void testsMain() {
         return mutableData;
       });
       expect(transactionResult.committed, true);
-      expect(transactionResult.dataSnapshot.value > value, true);
+      expect(transactionResult.dataSnapshot!.value > value, true);
+    });
+
+    test('DataSnapshot supports null childKeys for maps', () async {
+      // Regression test for https://github.com/FirebaseExtended/flutterfire/issues/6002
+
+      final ref = FirebaseDatabase.instance.reference().child('counter');
+
+      final transactionResult = await ref.runTransaction((mutableData) async {
+        mutableData.value = {'v': 'vala'};
+        return mutableData;
+      });
+
+      expect(transactionResult.committed, true);
+      expect(
+        transactionResult.dataSnapshot!.value,
+        {'v': 'vala'},
+      );
     });
 
     test('setPersistenceCacheSizeBytes Integer', () async {

--- a/packages/firebase_database/firebase_database/example/test_driver/firebase_database_e2e.dart
+++ b/packages/firebase_database/firebase_database/example/test_driver/firebase_database_e2e.dart
@@ -51,6 +51,9 @@ void testsMain() {
     test('runTransaction', () async {
       final FirebaseDatabase database = FirebaseDatabase.instance;
       final DatabaseReference ref = database.reference().child('counter');
+
+      await ref.set(0);
+
       final DataSnapshot snapshot = await ref.once();
       final int value = snapshot.value ?? 0;
       final TransactionResult transactionResult =
@@ -58,6 +61,7 @@ void testsMain() {
         mutableData.value = (mutableData.value ?? 0) + 1;
         return mutableData;
       });
+
       expect(transactionResult.committed, true);
       expect(transactionResult.dataSnapshot!.value > value, true);
     });

--- a/packages/firebase_database/firebase_database/lib/src/database_reference.dart
+++ b/packages/firebase_database/firebase_database/lib/src/database_reference.dart
@@ -146,13 +146,13 @@ class DatabaseReference extends Query {
   /// Performs an optimistic-concurrency transactional update to the data at
   /// this Firebase Database location.
   Future<TransactionResult> runTransaction(
-      TransactionHandler transactionHandler,
-      {Duration timeout = const Duration(seconds: 5)}) async {
-    assert(timeout.inMilliseconds > 0,
-        'Transaction timeout must be more than 0 milliseconds.');
-
-    final Completer<TransactionResult> completer =
-        Completer<TransactionResult>();
+    TransactionHandler transactionHandler, {
+    Duration timeout = const Duration(seconds: 5),
+  }) async {
+    assert(
+      timeout.inMilliseconds > 0,
+      'Transaction timeout must be more than 0 milliseconds.',
+    );
 
     final int transactionKey = FirebaseDatabase._transactions.isEmpty
         ? 0
@@ -165,7 +165,7 @@ class DatabaseReference extends Query {
           map['error'] != null ? DatabaseError._(map['error']) : null;
       final bool committed = map['committed'];
       final DataSnapshot? dataSnapshot = map['snapshot'] != null
-          ? DataSnapshot._fromJson(map['snapshot'], map['childKeys'])
+          ? DataSnapshot._fromJson(map['snapshot'], null)
           : null;
 
       FirebaseDatabase._transactions.remove(transactionKey);
@@ -173,21 +173,19 @@ class DatabaseReference extends Query {
       return TransactionResult._(databaseError, committed, dataSnapshot);
     }
 
-    // TODO(rrousselGit) refactor to async/await
-    // TODO(rrousselGit) what if invokeMethod fails?
-    // ignore: unawaited_futures
-    _database._channel.invokeMethod<void>(
-        'DatabaseReference#runTransaction', <String, dynamic>{
-      'app': _database.app?.name,
-      'databaseURL': _database.databaseURL,
-      'path': path,
-      'transactionKey': transactionKey,
-      'transactionTimeout': timeout.inMilliseconds
-    }).then((dynamic response) {
-      completer.complete(toTransactionResult(response));
-    });
+    final response =
+        await _database._channel.invokeMethod<Map<Object?, Object?>>(
+      'DatabaseReference#runTransaction',
+      <String, dynamic>{
+        'app': _database.app?.name,
+        'databaseURL': _database.databaseURL,
+        'path': path,
+        'transactionKey': transactionKey,
+        'transactionTimeout': timeout.inMilliseconds
+      },
+    );
 
-    return completer.future;
+    return toTransactionResult(response!);
   }
 
   OnDisconnect onDisconnect() {

--- a/packages/firebase_database/firebase_database/lib/src/event.dart
+++ b/packages/firebase_database/firebase_database/lib/src/event.dart
@@ -38,8 +38,8 @@ class DataSnapshot {
     Object? dataValue = _data['value'];
     Object? value;
 
-    if (dataValue is Map<Object?, Object?>) {
-      value = {for (final key in childKeys!) key: dataValue[key]};
+    if (dataValue is Map<Object?, Object?> && childKeys != null) {
+      value = {for (final key in childKeys) key: dataValue[key]};
     } else if (dataValue is List<Object?>) {
       value = childKeys!
           .map((key) => dataValue[int.parse(key! as String)])

